Sample rate detection

sukibaby opened the pull request:
Windows:
 -  100% done and working

Linux:
 - PulseAudio auto detection is not implemented, but I have established the framework for conditional behavior depending on whether the game is built with PulseAudio support.
 - Since PulseAudio still uses something like ALSA, the fallback of 48000 Hz for Unix OS's is based on the most likely value to prevent ALSA from needing to resample. It is also the default OSS sample rate.
 - I would leave it up to others to implement the PulseAudio solution.
 - The original `pactl`-based method is [here](https://github.com/itgmania/itgmania/pull/942/files#diff-2031301a15cc510b2303cf62a8a41016d266d37096cff09ebbd96f181076d58a), which works, but I don't like it.

Mac:
 -  hasn't been tested
 -  should be working 

------------------

More details below:

<details>

# Automatic sample rate detection

Here is the code that can detect the default audio device's sample rate and set the game's sample rate accordingly.

It is fully working, but not tested on Mac due to not having access to a Mac.

![image](https://github.com/user-attachments/assets/8c5c7f13-fd1c-4cf1-bab4-5c89076b77e1)
_Sample rate detection working while SoundPreferredSampleRate is set to 0._

# Summary of changes

First of all I implemented a new virtual function within ArchHooks. This is because the sample rate detection is done thru the OS rather than the sound driver. The function is called `DetermineSampleRate()`.

Second I changed the constexpr `kFallbackSampleRate` to an atomic `int`. This is because we want one singular reference that is global state, but we want to be able to overwrite it too. This allows existing logic using `kFallbackSampleRate` to remain in place while allowing for needed behavior.

RageSoundDriver gets a new static function,  `TryToGetSampleRate()`, which reads the value of `DetermineSampleRate()`.  First, it determines if there was any error, and clamps the value to a reasonable range, and logs any error. It also updates the atomic variable with the clamped value.

Finally, RageSoundDriver gets a new variable `validated_samplerate` which is populated and its value logged just before the loop to attempt to initialize an audio driver.

At that point the audio driver initializes as expected with a sample rate that matches the OS setting.

# Mac

The Mac implementation is untested.
Testing on Mac would be greatly appreciated.
</details>

sukibaby commented:
Setting as draft while testing is being undergone

sergioperez commented on src/arch/Sound/ALSA9Helpers.cpp:
```diff
@@ -231,7 +231,7 @@ RString Alsa9Buf::Init( int channels_,
 	preferred_chunksize = iChunkSize;
 	if( iSampleRate == 0 )
```
If I understood correctly, this is an uncommon scenario, as, when Alsa is used directly, ITGMania is able to set the sample rate of the hardware device.

Do you think it would be worth to print an error log in this scenario? At least, to be able to detect in runtime that something is wrong with the ITGMania/system configuration

PS: Thanks for this PR :D

sukibaby replied:
Hi,

Thanks for the review.

To be honest, I'm not really sure why Alsa9Buf::Init is written this way.  I was simply updating references to `kFallbackSampleRate`.


BTW, Linux sample rate detection currently only works if `pactl` is available, if you have  suggestions on how to do  it better I would definitely be open.

I will probably replace the implementation in this PR with a more aggressive one that totally disables the SoundPreferredSampleRate setting, as well as removing any possibility for more than one sample rate to be defined.

sergioperez replied:
Sure, thanks! I will try to help with the resources I have :)

As far as I understood, when ALSA is used, ITGMania takes the audio card exclusively for itself, and the system will "not have a sample rate" until the game starts, as there would not be a shared audio server running:

```
# cat /proc/asound/card0/pcm0p/sub0/hw_params
closed
```

Once we run ITGMania, setting `SoundPreferredSampleRate=16000`, we get the following logs:

```
00:00.045: RageSoundDriver: Using fallback sample rate 44100
00:00.057: ALSA: Mixing at 16000hz
```

The first one appears because there is no `pactl` in the system, as this is a alsa-only environment (no desktop, just Debian+LightDM+ITGMania).

However, given the second log, the game seems to successfully make ALSA sample at 16000Hz.

But, if I try to see what is going on in the card, I get a bit more confused:

```
# cat /proc/asound/card0/pcm0p/sub0/hw_params      
access: MMAP_INTERLEAVED
format: S16_LE
subformat: STD
channels: 2
rate: 48000 (48000/1)
period_size: 1024
buffer_size: 16384
```

I will try to find a solution to detect the available hardware sample rates in this scenario, and to understand better this information. But, my knowledge about audio in Linux is very limited. Maybe someone reads this conversation and gives us some light about it :)

(I am using a VM, I will use some actual hardware to try)

sukibaby replied:
Thank you for the insight. I plan to disable `SoundPreferredSampleRate` entirely,  so anything to ask ALSA for its default sample rate is perfect.

The problem is,  I'm looking to do it **before** initializing the audio driver, which seems to be the tricky part with ALSA  (or maybe I just don't know how to do it?)

sergioperez replied:
I think I have some findings (if it's okay, I will keep adding info over time here)

-  If I do not set a SoundDevice or I set it to `default`/`sysdefault`, as:

```
SoundDevice= 
SoundDrivers=ALSA-sw
SoundPreferredSampleRate=16000
```

Then, clearly, the sampling rate of the audio I hear is 16000Hz, but, the hardware sampling rate reports to be 48kHz:
```
$ cat /proc/asound/card0/stream0 
(omitted info)
    Momentary freq = 48000 Hz (0x30.0000)
```

- **But**, if I set it to use `hw:0,0` `plughw:0,0` as the sound device 

```
SoundDevice=hw:0,0 (or plughw:0,0)
SoundDrivers=ALSA-sw
SoundPreferredSampleRate=16000
```

Then, ITGMania is what sets the sample rate in the card:
```
$ cat /proc/asound/card0/stream0 
(omitted info)
Momentary freq = 16000 Hz (0x10.0000)
```

Hence, in these cases, it would not be required to perform any detection, as, if the card is set to be initialized with a 16kHz sample rate from ITGMania, that will be what the hardware will use. (Or actually, the "nearest" sample rate)

- Now, another confusing behavior. When the SoundDevice is set to dmix, and the sample rate to 16kHz:
```
SoundDevice=dmix
SoundDrivers=ALSA-sw
SoundPreferredSampleRate=16000
```
We find out that ITGMania sets its sample rate to 48000Hz, matching the output that initialized the sound card. As far as I understood, this is an ALSA plugin to let different applications output audio at the same time. Maybe in this scenario, `dsnd_pcm_hw_params_set_rate_near` will just provide what is configured as the default in the card.

Then, as a brief, I find the following scenarios, if I set a 16000Hz sample ratio:

- Device: plughw,hw => ITGMania defines the sample rate. HW + game goes 16kHz. No detection needed.


- Device: dmix => ITGMania will output at the default (48000Hz) sample rate set in `/usr/share/alsa/alsa.conf`. All goes 48kHz, regardless of `SoundPreferredSampleRate`. I guess it is not worth the effort to detect anything.


- Device: empty or default/sysdefault => The device will be initialized with a 48000Hz sample rate. But ITGMania will send 16000Hz audio. I'm not sure about how to approach this scenario.

sergioperez replied:
After testing a bit more, I think the scenarios with ALSA are:



| SoundDevice     | First action                                                | Second action                                                                 |
| ----------------------- | ----------------------------------------------------- | -------------------------------------------------------------------------|
| (empty/default) | dmix initializes the card to 48000Hz   | The `plug` plugin resamples the audio to 48000Hz |
| dmix                    | dmix initializes the card to 48000Hz   |  dsnd_pcm_hw_params_set_rate_near will always set 48000Hz, as it will appear as the only available rate |
| hw:0,0                  | ITGMania will initialize the card to the nearest sample rate to the one we specify | None |

Note1: This is the behaviour with the default (`/usr/share/alsa/alsa.conf`) settings from a default Debian 12 install (no `.asoundrc`)
Note2: I might not be technically accurate in some parts of what I am writing here.

How to test what is going on with the `default` audio device:
```
# play a .wav audio
aplay -v /usr/share/sounds/alsa/Noise.wav
# resample the wav file to 16000Hz
sox -r 16000 /usr/share/sounds/alsa/Noise.wav noise.wav
# play it
aplay -v noise.wav 
```

Maybe, an scenario that would work to detect the sample rate of a `default` audio device would be:
1. Initializing the audio device
2. Reading the hardware sample rate
3. Stopping the audio device
4. Re-initializing the audio device

But, somehow, this gives a meaning to have `SoundPreferredSampleRate`: When an audio device is used directly and without plugins (`hw:X,Y`), there is not a default sample rate. ITGMania decides it, and both the game and the audio card will sample as decided.

sukibaby replied:
Thanks again so much for this!   I'll definitely shout you out in the commit for your help with this.  I never really dived into ALSA before and ended up realizing it operates at a lower level than the other audio protocols which ITGm has drivers for.

I've been thinking it over for a while and I'm starting to think that if a system has ALSA, but not Pulseaudio or Pipewire,  then it's likely intentional or the user is on a seriously old OS.

In either case I think I can do a similar approach to the `pactl` logic in case the user has Pipewire + ALSA but no Pulseaudio. If there's no Pulseaudio and no Pipewire but there is ALSA I think it would be reasonable to declare 48000, especially if that's what dmix initializes the card to.

sergioperez replied:
Thanks! I think that would be a good start, because it would work for the default settings of almost all distros :)

And for example, if a system is set to use only hw:0,0@44100 (no plug, no dmix), the ALSA API call to set the sample rate at 48000 will just set it to 44100

But, I can also review options to implement the four steps of the previous comment. So if someone did set dmix to 44100, ITGMania would be set to 44100 - I an add some code snippets here, or a PR in the future :)

About the use cases, one example using only ALSA: I have an Ansible Playbook that sets up an "ITGMania arcade machine" in any system with a Debian 12 minimal installation:
https://github.com/sergioperez/stepmania-cabinet-tools

With it, I try to reduce to the minimum the amount of software installed on the system, and also, to get the less posible amount of layers between the game and the audio hardware device.

Although, I am not sure about if it's preferable or not to keep a parameter to let players using ALSA set the frequency when using a HW device. In theory, at some point in the future, this would keep the option open to select the audio device from the theme settings

sergioperez replied:
I've done some small tests again!

It would be possible to check the default dmix sample rate in a system:
```
        // Check dmix sample rate
        snd_config_t *value;
        long dmix_rate;
        snd_config_update();
        if(snd_config_search(snd_config, "defaults.pcm.dmix.rate", &value) != 0) {
                samplerate = g_FallbackSampleRate.load();
                LOG->Info( "Failed to fetch dmix sample rate - Falling back to %i", g_FallbackSampleRate.load() );
        } else {
                snd_config_get_integer(value, &rate);
        }
        // End of check dmix sample rate
```

This would work if:
1. The audio is initialized
2. The sample rate is checked
3. The audio is re-initialized with the correct sample rate

However, I did not find a way to detect if the current audio stream is using plugins or directly the hardware device. Maybe a simple way to achieve this, is to check if `PREFSMAN->m_iSoundDevice.Get()` is empty or `default` -> If `hw:X,Y` is set as the audio device, the dmix sample rate does not matter for us, as the game initializes the hardware with any desired sample rate.
Edit: This might also not be a good idea, as the user can override the default device for a hw device in .asoundrc

For dmix:
- The configured system sample rate can be checked
- But it is also safe to assume that the audio card will be started at 48000Hz

For direct hw access:
- If 48000Hz is set as the sample rate, but the card only supports 16000Hz, the hw and the game will use 16000Hz (any value should be just safe)
- Maybe it's worth keeping `SoundPreferredSampleRate` just for ALSA with direct hardware access, even if it's not used by the other drivers? - Not sure about if there is a scenario were we would prefer to have a sample rate different from 44100 or 48000

Unless at the moment, to keep things simple, everything is just set to 48000Hz.

Edit: Or to set 48000 or alsa-dmix-autodetected, unless there is a value in SoundPreferredSampleRate

sukibaby replied:
Thanks again @sergioperez ! this is awesome.

I did some testing on my Linux machine (Debian 12 + KDE) by removing PulseAudio entirely (PipeWire isn't there by default).

As expected, we fall back to 44100 hz since it can't get a result from `pactl`, but I think I'd rather fall back to 48000 if we seem to be running on an ALSA-only system, because my findings imply that even though the log file confirms ITGmania initialized the ALSA driver at 44100 Hz, that ALSA is actually running at 48000 Hz:

```
$ cat /proc/asound/card0/pcm0p/sub0/hw_params
access: MMAP_INTERLEAVED
format: S32_LE
subformat: STD
channels: 2
rate: 48000 (48000/1)
period_size: 1024
buffer_size: 16384
```

So I think I'm going to fall back to 48000 rather than 44100 if we seem to be on an ALSA only system. 

(We're looking to minimize or eliminate resampling happening outside of the game's control, so if under the hood it's negotiating 48000 Hz with the audio device and handling the resampling itself, that's what I want to prevent as much as possible).

sukibaby commented:
This PR is now ready for review, with many thanks to @sergioperez for the assistance with correctly handling ALSA (he may also have a follow up PR for ALSA changes which would be outside of the scope of this PR).

Windows and Linux are fully tested.   I personally tested PulseAudio + ALSA, JACK + ALSA, and ALSA by itself on Linux.

Mac still needs testing, but it should be good to go if Mac is all good.

sukibaby commented:
@teejusb  i do want to say,  i'm not a big fan of the "see if pactl exists" based approach for unix.  to get around this i would imagine i would ~~bring in `pulse/pulseaudio.h` and add a callback to get it, and fall back to 48000 for ALSA/OSS if it doesn't detect any pulseaudio on the host PC.  what are your thoughts on this matter?~~

edit: I tried some stuff and I'm actually not sure how to get this default sample rate via the API without initializing the driver, but the whole point is to get the sample rate **before** initializing the driver. So I'm not sure how to best handle this.  I really hate the idea of querying `pactl` via `FILE`, but it was able to do this. At the moment, I'm not sure what's best for Linux. Maybe you have an idea here?

sukibaby committed:
Replace kFallbackSampleRate with g_FallbackSampleRate

sukibaby committed:
Let ITGmania know about CMake's HAS_PULSE macro

sukibaby committed:
Determine the sample rate before initializing any audio driver

sukibaby committed:
Misc reversions of includes to pre-kFallbackSampleRate status

sukibaby committed:
Implement ability to determine sample rate

sukibaby committed:
Placeholder for unfinished PulseAudio sample rate detection

The console based method of asking pactl what the sample rate is works, but is hideous.  I've set this up so there is already a framework to conditionally behave differently depending on whether or not the game was compiled with PulseAudio support. I just haven't implemented the PulseAudio sample rate detection via the API.
